### ISSUE TITLE
Build an all locales html file

### ISSUE THREAD

**Issue by mattpen** [phetsims/chipper#563]
We need to setup a process for building a html file with all locales, test it, and add it to the production build process.

**Comment by mattpen** [comment]
From dev meeting:

> MP: Formulate a plan for handling sim translations within the mobile apps
>Where are we in terms of implementing a single-file sim with all translations included?  
>How would we switch locales using such a file, query parameters?
>MKauz: This was talked about in regards to the installer-builder. Perhaps this new use case bumps up the priority: https://github.com/phetsims/installer-builder/issues/54
>JO: This would require a maintenance release for chipper for all sims
>AP: What is the best way to proceed with this. One sim (chains) first?

**Comment by jbphet** [comment]
Plan discussed in 4/13/2017 dev meeting:

+ Test the generation of the "all" html build (the capability has been implemented, but hasn't been used or tested much)
+ Identify the published simulations that don't support the generation of the "all" build and do maintenance releases for each
+ Add building of the "all" file to the build server
+ Figure out the meta data necessary for the iOS and Android apps to determine when updates have occured to a sim for the locale(s) that the user has identified as important and add it to the web site
+ Modify the iOS and Android apps to use the "all" files instead of the "en" files and to specify the locales in which they are interested
+ Add the capability to the iOS and Android apps to download updates for sims based on new translations being available

**Comment by jonathanolson** [comment]
> Test the generation of the "all" html build (the capability has been implemented, but hasn't been used or tested much)

Tested in master (molarity). Builds a combined file, `phet.chipper.strings` has keys: en, ar, bs, da, de, el, es_CO, es_ES, es_MX, es_PE, es, et, eu, fa, fi, fr, gl, hu, in, it, iw, ja, ka, kk, km, ko, lv, mk, mr, nb, nl, pl, pt_BR, ru, si, sk, sr, sv, ta, th, tk, vi, zh_CN, zh_TW. Spot-checked and the `?locale=` query parameter seemed to work nicely.

**Comment by jonathanolson** [comment]
> Identify the published simulations that don't support the generation of the "all" build and do maintenance releases for each

Will presumably wait on other chipper 2.0 issues to see if we can combine maintenance releases.

**Comment by jonathanolson** [comment]
Original issue: https://github.com/phetsims/chipper/issues/509

**Comment by jonathanolson** [comment]
Using the new shaCheck (determines what released simulations have given SHAs included) and through checking automated builds, I've determined that the following sims will need maintenance releases for this:

acid-base-solutions
area-builder
arithmetic
atomic-interactions
balancing-act
balancing-chemical-equations
beers-law-lab
bending-light
charges-and-fields
color-vision
concentration
energy-skate-park-basics
faradays-law
forces-and-motion-basics
fraction-matcher
friction
function-builder
hookes-law
isotopes-and-atomic-mass
least-squares-regression
molecule-shapes
molecule-shapes-basics
molecules-and-light
neuron
ohms-law
ph-scale
ph-scale-basics
reactants-products-and-leftovers
resistance-in-a-wire
rutherford-scattering
trig-tour
under-pressure
wave-on-a-string

as the following sims already have the _all.html support:

balloons-and-static-electricity
build-an-atom
capacitor-lab-basics
circuit-construction-kit-dc
circuit-construction-kit-dc-virtual-lab
expression-exchange
function-builder-basics
graphing-lines
graphing-slope-intercept
gravity-and-orbits
gravity-force-lab
john-travoltage
make-a-ten
molarity
molecule-polarity
pendulum-lab
plinko-probability
projectile-motion
proportion-playground
states-of-matter
states-of-matter-basics
unit-rates

**Comment by jonathanolson** [comment]
The patches are semi-involved so far (6-7 files across chipper/joist so far), and I've had to make 2 moderately different patches so far.

Due to that, I think two types of review are going to be needed.

- I'll want code reviews on all of the independent patches I'll be writing, as I'm basically reimplementing the feature on different codebases. @jbphet and/or @mattpen, whoever feels most familiar with how strings are handled in the sim (and how build-server builds using the --locales parameter) is who I'd recommend.
- I think proper QA testing for this is going to require someone being able to check out release branches (new `grunt checkout-release --repo={sim}` in perennial), run various grunt build commands, and manually verify that the _en.html, _{{LOCALE}}.htmls, and _all.html work and have the expected global for phet.chipper.strings. @phet-steele, would QA be able to handle this (I can help with setup and explanation), or would we need a developer for this? (I'm estimating 5-10mins/sim, 3-5 hours total)

**Comment by jonathanolson** [comment]
The first two sims I patched turned out to need the only two different implementations. For review:

- Chipper changes, https://github.com/phetsims/chipper/commit/ed5fadf1df25d856ea14ae440a68fa1e961ee28a and https://github.com/phetsims/chipper/commit/eb55232520e8ad23bbf68ad14817cd1f8c7ae752
- (minimal) Joist changes, 1d73ca2132952c6678a39acbdf4c173030691b94 and 88184ee60f63945e6529699f86637a67a278979f

@jbphet or @mattpen, can you look into these changes (mainly that the two chipper patches should do equivalent things)?

**Comment by jonathanolson** [comment]
RCs pushed out, created https://github.com/phetsims/QA/issues/72 for tracking testing. Will track code review in this issue.

**Comment by jbphet** [comment]
I've reviewed the two chipper patches, i.e. https://github.com/phetsims/chipper/commit/ed5fadf1df25d856ea14ae440a68fa1e961ee28a for acid-base-solutions 1.2.x and https://github.com/phetsims/chipper/commit/eb55232520e8ad23bbf68ad14817cd1f8c7ae752 for atomic-interactions-1.0.  Both look reasonable to me, but are hard to review without spending a fair amount of time making sure that the changes match up with the state of the code when these branches are made.  The real test will be to verify that these branches all the generation of the _all.html files and that those file function correctly.  So, review passed, but QA need to be thorough if we want to have a high degree of confidence that this was done correctly.

**Comment by jbphet** [comment]
Assigning back to @jonathanolson, since I'm not sure if he wanted to pass this on to QA or do something else next.

**Comment by jonathanolson** [comment]
Note to self: Don't run the RIAW deploy, a newer version was published.

**Comment by jonathanolson** [comment]
Maintenance releases completed (following list has the deployment confirmation HTMLs):

Deployed:


- [ ] https://phet.colorado.edu/sims/html/acid-base-solutions/1.2.11/acid-base-solutions_en.html
- [ ] https://phet.colorado.edu/sims/html/area-builder/1.1.6/area-builder_en.html
- [ ] https://phet.colorado.edu/sims/html/arithmetic/1.0.10/arithmetic_en.html
- [ ] https://phet.colorado.edu/sims/html/atomic-interactions/1.0.4/atomic-interactions_en.html
- [ ] https://phet.colorado.edu/sims/html/balancing-act/1.1.10/balancing-act_en.html
- [ ] https://phet.colorado.edu/sims/html/balancing-chemical-equations/1.1.10/balancing-chemical-equations_en.html
- [ ] https://phet.colorado.edu/sims/html/beers-law-lab/1.4.4/beers-law-lab_en.html
- [ ] https://phet.colorado.edu/sims/html/bending-light/1.1.7/bending-light_en.html
- [ ] https://phet.colorado.edu/sims/html/charges-and-fields/1.0.4/charges-and-fields_en.html
- [ ] https://phet.colorado.edu/sims/html/color-vision/1.1.10/color-vision_en.html
- [ ] https://phet.colorado.edu/sims/html/concentration/1.3.6/concentration_en.html
- [ ] https://phet.colorado.edu/sims/html/energy-skate-park-basics/1.1.6/energy-skate-park-basics_en.html
- [ ] https://phet.colorado.edu/sims/html/faradays-law/1.1.10/faradays-law_en.html
- [ ] https://phet.colorado.edu/sims/html/forces-and-motion-basics/2.1.5/forces-and-motion-basics_en.html
- [ ] https://phet.colorado.edu/sims/html/fraction-matcher/1.1.8/fraction-matcher_en.html
- [ ] https://phet.colorado.edu/sims/html/friction/1.3.4/friction_en.html
- [ ] https://phet.colorado.edu/sims/html/function-builder/1.0.9/function-builder_en.html
- [ ] https://phet.colorado.edu/sims/html/hookes-law/1.0.10/hookes-law_en.html
- [ ] https://phet.colorado.edu/sims/html/isotopes-and-atomic-mass/1.0.9/isotopes-and-atomic-mass_en.html
- [ ] https://phet.colorado.edu/sims/html/least-squares-regression/1.1.7/least-squares-regression_en.html
- [ ] https://phet.colorado.edu/sims/html/molecule-shapes/1.1.11/molecule-shapes_en.html
- [ ] https://phet.colorado.edu/sims/html/molecule-shapes-basics/1.1.9/molecule-shapes-basics_en.html
- [ ] https://phet.colorado.edu/sims/html/molecules-and-light/1.1.16/molecules-and-light_en.html
- [ ] https://phet.colorado.edu/sims/html/neuron/1.1.4/neuron_en.html
- [ ] https://phet.colorado.edu/sims/html/ohms-law/1.3.8/ohms-law_en.html
- [ ] https://phet.colorado.edu/sims/html/ph-scale/1.2.11/ph-scale_en.html
- [ ] https://phet.colorado.edu/sims/html/ph-scale-basics/1.2.11/ph-scale-basics_en.html
- [ ] https://phet.colorado.edu/sims/html/reactants-products-and-leftovers/1.1.10/reactants-products-and-leftovers_en.html
- [ ] https://phet.colorado.edu/sims/html/rutherford-scattering/1.0.7/rutherford-scattering_en.html
- [ ] https://phet.colorado.edu/sims/html/trig-tour/1.0.9/trig-tour_en.html
- [ ] https://phet.colorado.edu/sims/html/under-pressure/1.1.5/under-pressure_en.html
- [ ] https://phet.colorado.edu/sims/html/wave-on-a-string/1.1.9/wave-on-a-string_en.html

**Comment by jonathanolson** [comment]
Verified RIAW also, no patch needed.

Assigning to @mattpen for build-server work to enable the --allHTML flag.

**Comment by mattpen** [comment]
This should be handled, but the build-server is not ready to test yet.

**Comment by mattpen** [comment]
This is working now in the build-server.  See https://www.colorado.edu/physics/phet/dev/html/chains/1.7.0-rc.20/phet/chains_all_phet.html

**Comment by mattpen** [comment]
@jonathanolson - What are the next steps? Is it ok to close this issue?

**Comment by jonathanolson** [comment]
Presumably you'd want to rebuild all production sims so that the file exists (and will continue to exist when rosetta updates things).

**Comment by mattpen** [comment]
We should discuss the plan for rebuilding all sims to generate the _all.html files

**Comment by pixelzoom** [comment]
1/18/18 dev meeting notes:
• Rosetta doesn't commit anything if translation hasn't changed, but it does do a build. @mattpen verified using chains.
• Create a new issue to have @DianaTavares resubmit her translations (for all sims) to cause a rebuild. Assign to @oliver-phet and @DianaTavares.

**Comment by oliver-phet** [comment]
@mattpen Diana resubmitted all of her HTML5 sim translations. I spotted checked a few of the sim directories and the _all.html file has been generated. Is there anything else needed?

**Comment by jbphet** [comment]
@mattpen - I'm trying to keep an eye on this issue, but please also notify me via Slack when you've confirmed that all of the `_all.html` files are in place so that I can work towards getting the installer builder to use them.

**Comment by mattpen** [comment]
@jbphet - The following sims have _all.html files available.  I ran a basic http GET request for all sims in `/perennial/data/active-sims` and these returned 200 OK.  There are 58 sims here which I think means this is everything (56 published + example-sim & chains).  I have not tested all locales in all files.

```
acid-base-solutions
area-builder
arithmetic
atomic-interactions
balancing-act
balancing-chemical-equations
balloons-and-static-electricity
beers-law-lab
bending-light
build-an-atom
capacitor-lab-basics
chains
charges-and-fields
circuit-construction-kit-dc
circuit-construction-kit-dc-virtual-lab
color-vision
concentration
energy-skate-park-basics
example-sim
expression-exchange
faradays-law
forces-and-motion-basics
fraction-matcher
friction
function-builder
function-builder-basics
gene-expression-essentials
graphing-lines
graphing-slope-intercept
gravity-and-orbits
gravity-force-lab
hookes-law
isotopes-and-atomic-mass
john-travoltage
least-squares-regression
make-a-ten
molarity
molecules-and-light
molecule-polarity
molecule-shapes
molecule-shapes-basics
neuron
ohms-law
pendulum-lab
ph-scale
ph-scale-basics
plinko-probability
projectile-motion
proportion-playground
reactants-products-and-leftovers
resistance-in-a-wire
rutherford-scattering
states-of-matter
states-of-matter-basics
trig-tour
under-pressure
unit-rates
wave-on-a-string
```

**Comment by jbphet** [comment]
Thanks, @mattpen.  I'll test this.  If anyone is interested, the related issue in the installer-builder repo is https://github.com/phetsims/installer-builder/issues/54.

**Comment by mattpen** [comment]
Work for this issue is complete. Closing because installer-builder work is tracked separately.
